src/Solvers/multigrid_solver.jl

```diff
@@ -1,22 +1,45 @@
 using Oceananigans.Architectures: architecture
-using Oceananigans.Grids: interior_parent_indices
-using Statistics: norm, dot
 using LinearAlgebra
 using AlgebraicMultigrid: _solve!, init, RugeStubenAMG
-using Oceananigans.Operators: volume, Δyᶠᶜᵃ, Δyᶜᶠᵃ, Δyᶜᶜᵃ, Δxᶠᶜᵃ, Δxᶜᶠᵃ, Δxᶜᶜᵃ, Δyᵃᶜᵃ, Δxᶜᵃᵃ, Δzᵃᵃᶠ, Δzᵃᵃᶜ, ∇²ᶜᶜᶜ
+using CUDA
+using AMGX
 
 import Oceananigans.Architectures: architecture
 
-mutable struct MultigridSolver{A, G, L, T, M, F}
-               architecture :: A
-                       grid :: G
-                     matrix :: L
-                     abstol :: T
-                     reltol :: T
-                    maxiter :: Int
-              amg_algorithm :: M
-                    x_array :: F
-                    b_array :: F
+abstract type MultigridSolver{A, G, L, T, F} end
+
+mutable struct MultigridCPUSolver{A, G, L, T, F, M} <: MultigridSolver{A, G, L, T, F}
+                  architecture :: A
+                          grid :: G
+                        matrix :: L
+                        abstol :: T
+                        reltol :: T
+                       maxiter :: Int
+                       x_array :: F
+                       b_array :: F
+                 amg_algorithm :: M
+end
+
+mutable struct MultigridGPUSolver{A, G, L, T, F, S} <: MultigridSolver{A, G, L, T, F}
```

Comment:
it seems that `MultiGridCPUSolver` and `MultiGridGPUSolver` have the same fields. 

Can you just define 
  ```
mutable struct MultigridSolver{A, G, L, T, F, M} 
                  architecture :: A
                          grid :: G
                        matrix :: L
                        abstol :: T
                        reltol :: T
                       maxiter :: Int
                       x_array :: F
                       b_array :: F
                 amg_algorithm :: M
end

const MultigridCPUSolver = MultigridSolver{<:CPU}
const MultigridGPUSolver = MultigridSolver{<:GPU}
```

Reply:
Oh, they don't. But your suggestion might come handy indeed! I had forgotten that this `MultigridSolver{<:CPU}` would work!